Initial doc updates (#188)
Azure-Samples/azure-search-openai-demo-csharp: README.md

> # ChatGPT + Enterprise data with Azure OpenAI and Cognitive Search (.NET)

Reviewer: Can we call it "ChatGPT"? I get why we'd do it (from an associated branding) but OpenAI has filed a trademark request for ChatGPT (which seems to have been denied and they are appealing), so it's possibly in a legal grey area at the moment.

Author: We're following similar naming as here but good question. 

https://github.com/Azure-Samples/azure-search-openai-demo

Tracking here #196

![Chat screen](docs/chatscreen.png)

## Application architecture

> - **User interface** - The application’s chat interface is a [Blazor WebAssembly](https://learn.microsoft.com/aspnet/core/blazor/) static web application. This interface is what accepts user queries, routes request to the application backend, and displays generated responses.

Reviewer: Is "static web application" the best descriptor here? Should we state it's a "Blazor Web Assembly application" instead?

Author: Good call. It's technically hosted inside the Web service today.

- **Backend** - The application backend is an [ASP.NET Core Minimal API](https://learn.microsoft.com/aspnet/core/fundamentals/minimal-apis/overview). The backend hosts the Blazor static web application and what orchestrates the interactions among the different services. Services used in this application include:
   - [**Azure Cognitive Search**](https://learn.microsoft.com/azure/search/search-what-is-azure-search) – indexes documents from the data stored in an Azure Storage Account. This makes the documents searchable using [vector search](https://learn.microsoft.com/azure/search/search-get-started-vector) capabilities. 

>    - [**Azure OpenAI Service**](https://learn.microsoft.com/azure/ai-services/openai/overview) – provides the ChatGPT models to generate responses. The Additionally, [Semantic Kernel](https://learn.microsoft.com/semantic-kernel/whatissk) is used in conjunction with the Azure OpenAI Service to orchestrate the more complex AI workflows.

Reviewer: ```suggestion
   - [**Azure OpenAI Service**](https://learn.microsoft.com/azure/ai-services/openai/overview) – provides the Large Language Models to generate responses. The Additionally, [Semantic Kernel](https://learn.microsoft.com/semantic-kernel/whatissk) is used in conjunction with the Azure OpenAI Service to orchestrate the more complex AI workflows.
```

- **Azure account permissions** - Your Azure Account must have `Microsoft.Authorization/roleAssignments/write` permissions, such as [User Access Administrator](https://learn.microsoft.com/azure/role-based-access-control/built-in-roles#user-access-administrator) or [Owner](https://learn.microsoft.com/azure/role-based-access-control/built-in-roles#owner).

> **Warning**<br>
> By default this sample will create an Azure Container App, and Azure Cognitive Search resource that have a monthly cost, as well as Form Recognizer resource that has cost per document page. You can switch them to free versions of each of them if you want to avoid this cost by changing the parameters file under the infra folder (though there are some limits to consider; for example, you can have up to 1 free Cognitive Search resource per subscription, and the free Form Recognizer resource only analyzes the first 2 pages of each document.)

### Cost estimation

> Pricing varies per region and usage, so it isn't possible to predict exact costs for your usage. However, you can try the [Azure pricing calculator](https://azure.microsoft.com/en-us/pricing/calculator/) for the resources below:

Reviewer: ```suggestion
Pricing varies per region and usage, so it isn't possible to predict exact costs for your usage. However, you can try the [Azure pricing calculator](https://azure.microsoft.com/pricing/calculator/) for the resources below:
```

Make sure to drop the local from URLs so people will be sent to the right page.

- [**Azure Container Apps**](https://azure.microsoft.com/pricing/details/container-apps/)

> - [**Azure OpenAI Service**](https://azure.microsoft.com/en-us/pricing/details/cognitive-services/openai-service/)

Reviewer: ```suggestion
- [**Azure OpenAI Service**](https://azure.microsoft.com/pricing/details/cognitive-services/openai-service/)
```
Make sure to drop the local from URLs so people will be sent to the right page.

- [**Azure Form Recognizer**](https://azure.microsoft.com/pricing/details/form-recognizer/)
- [**Azure Cognitive Search**](https://azure.microsoft.com/pricing/details/search/)
- [**Azure Blob Storage**](https://azure.microsoft.com/pricing/details/storage/blobs/)
- [**Azure Monitor**](https://azure.microsoft.com/pricing/details/monitor/)

### Project setup

> You have a few options for setting up this project. The easiest way to get started is GitHub Codespaces, since it will setup all the tools for you, but you can also set it up [locally]() if desired.

Reviewer: Missing a link here.

   - [**Azure Cognitive Search**](https://learn.microsoft.com/azure/search/search-what-is-azure-search) – indexes documents from the data stored in an Azure Storage Account. This makes the documents searchable using [vector search](https://learn.microsoft.com/azure/search/search-get-started-vector) capabilities. 
   - [**Azure OpenAI Service**](https://learn.microsoft.com/azure/ai-services/openai/overview) – provides the Large Language Models to generate responses. [Semantic Kernel](https://learn.microsoft.com/semantic-kernel/whatissk) is used in conjunction with the Azure OpenAI Service to orchestrate the more complex AI workflows.

## Getting Started

### Account Requirements

In order to deploy and run this example, you'll need

- **Azure Account** - If you're new to Azure, get an [Azure account for free](https://aka.ms/free) and you'll get some free Azure credits to get started.

> - **Azure subscription with access enabled for the Azure OpenAI service** - You can request access [here](https://aka.ms/oaiapply). You can also visit [here](https://azure.microsoft.com/free/cognitive-search/) to get some free Azure credits to get you started.

Reviewer: Not a deal breaker, but I learnt a bit of advice on how we can make more accessible pages by avoiding using `here` as link text, as if someone is using a screen reader that's not helpful. [This video](https://www.linkedin.com/posts/geektrainer_adhd-a11y-accessibility-activity-7090030184413736960-ZpYw?utm_source=share&utm_medium=member_desktop) covers it off nicely.

```suggestion
- **Azure subscription with access enabled for the Azure OpenAI service** - [You can request access](https://aka.ms/oaiapply). You can also visit [the Cognitive Search docs](https://azure.microsoft.com/free/cognitive-search/) to get some free Azure credits to get you started.
```

- [Docker](https://www.docker.com/products/docker-desktop/)
  - **Important**: Ensure Docker is running before running any `azd` provisioning / deployment commands.

Then, run the following commands to get the project on your local environment:

   1. Run `azd auth login`

>    1. Run `azd env new azure-search-openai-demo-csharp`

Reviewer: Don't you need to do `azd init` to bring the code down? Or is the expectation that they are going to `git clone` this repo rather than use it as an azd template?

Author: Good question. It depends. If you haven't cloned the repo, calling `azd init` does that for you. Otherwise you can take the approcah you suggested.

Run `azd down`

### Using the app

> - In Azure: navigate to the Azure Contaier App deployed by `azd`. The URL is printed out when `azd` completes (as "Endpoint"), or you can find it in the Azure portal.

Reviewer: ```suggestion
- In Azure: navigate to the Azure Container App deployed by `azd`. The URL is printed out when `azd` completes (as "Endpoint"), or you can find it in the Azure portal.
```

- When running locally, navigate to <http://localhost:7181> for the client app and <http://localhost:7181/swagger> for the Open API server page.

Once in the web app:

- On the **Voice Chat** page, select the voice settings dialog and configure text-to-speech preferences.
  - You can either type messages to interact with Blazor Clippy, or select the Speak toggle button to use speech-to-text as your input.
- Try different topics in **Chat** context. For chat, try follow up questions, clarifications, ask to simplify or elaborate on answer, etc.
- Explore citations and sources
- Click on the "settings" icon to try different options, tweak prompts, etc.

## Enabling optional features

### Enabling Application Insights

### Enabling authentication

> By default, the deployed Azure container app will have no authentication or access restrictions enabled, meaning anyone with routable network access to the container app can chat with your indexed data.  You can require authentication to your Azure Active Directory by following the [Add container app authentication](https://learn.microsoft.com/en-us/azure/container-apps/authentication-azure-active-directory) tutorial and set it up against the deployed container app.

Reviewer: ```suggestion
By default, the deployed Azure container app will have no authentication or access restrictions enabled, meaning anyone with routable network access to the container app can chat with your indexed data.  You can require authentication to your Azure Active Directory by following the [Add container app authentication](https://learn.microsoft.com/azure/container-apps/authentication-azure-active-directory) tutorial and set it up against the deployed container app.
```

Make sure to drop the local from URLs so people will be sent to the right page.

To then limit access to a specific set of users or groups, you can follow the steps from [Restrict your Azure AD app to a set of users](https://learn.microsoft.com/azure/active-directory/develop/howto-restrict-your-app-to-a-set-of-users) by changing "Assignment Required?" option under the Enterprise Application, and then assigning users/groups access.  Users not granted explicit access will receive the error message -AADSTS50105: Your administrator has configured the application <app_name> to block users unless they are specifically granted ('assigned') access to the application.-

> ## Productionizing

Reviewer: For the things we suggest down below, would it be better to make them parameters to Bicep (with the current as default) to allow overriding without diving into the Bicep?

Author: Probably

Author: Tracking here #197
